Integrate shared chunk cache (H5SC) package with upper layers of library

fortnern opened the pull request:
Also some updates to shared chunk cache interface, and other general fixes. Non functional currently since the shared chunk cache hasn't been implemented.

fortnern committed:
Add integration of shared chunk cache (H5SC) layer with top layer of
library.

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Fix spelling and cmake error (hopefully)

fortnern committed:
Merge branch 'feature/sparse_data' of github.com:fortnern/hdf5 into feature/sparse_data

fortnern committed:
Clean up shared chunk cache on failure in H5F_new()

mattjala commented on src/H5Dio.c:
```diff
@@ -436,21 +454,28 @@ H5D__read(size_t count, H5D_dset_io_info_t *dset_info)
 
+    /* Make shared chunk cache read call if appropriate */
+    if (any_scc) {
+        assert(H5F_get_shared_cache(dset_info[0].dset->oloc.file));
```
Does it make sense to check the first dataset's file for a shared cache, when the first dataset may not be using the shared chunk cache?

fortnern replied:
The shared chunk cache is global to the file so this code should find it even if that dataset doesn't use it. All datasets must share the same file.

fortnern committed:
Hopefully fix cmake

mattjala commented on src/H5Dio.c:
```diff
@@ -1003,30 +1055,33 @@ H5D__dset_ioinfo_init(H5D_t *dset, H5D_dset_io_info_t *dset_info, H5D_storage_t
-        dset_info->io_ops.single_write = H5D__scatgath_write;
-    } /* end else */
+    /* Only set layout ops if weŕe not using the shared chunk cache */
+    if (!dset->shared->layout.sc_ops) {
```
If the behavior from `H5D__read/writer()` of skipping IO info initialization for shared chunk cache datasets is going to be a consistent pattern, it might be better to just assert that the dataset isn't using the shared chunk cache. It would remove a layer of nesting, which is always nice.

fortnern replied:
This function is called even for datasets using the SCC. Do you mean we should avoid calling it for those datasets (and move up the setting of the dset and store parameters)?

mattjala replied:
Disregard my comment - I had IO info init and IO init confused.

mattjala commented on src/H5VLnative_dataset.c:
```diff
@@ -755,53 +756,75 @@ H5VL__native_dataset_optional(void *obj, H5VL_optional_args_t *args, hid_t dxpl_
+            if (dset->shared->layout.sc_ops) {
+                H5D_chunk_scc_udata_t udata;
+                size_t                buf_size = SIZE_MAX;
+                assert(0 && "set up the buf size from API when available. Either point directly into args "
```
Was this hardcoded assertion failure supposed to be left in for now?

fortnern replied:
Yes. We should be updating the API soon and will be able to get rid of this assertion.

fortnern commented on src/H5CX.c:
```diff
@@ -2900,6 +2900,40 @@ H5CX_set_mpio_actual_chunk_opt(H5D_mpio_actual_chunk_opt_mode_t mpio_actual_chun
     FUNC_LEAVE_NOAPI_VOID
 } /* end H5CX_set_mpio_actual_chunk_opt() */
 
+/*-------------------------------------------------------------------------
```
Added these since this should be more convenient than get/or/set. I think we could possibly use these in a few more places instead of tracking the property in a separate variable.

fortnern commented on src/H5Dchunk.c:
```diff
@@ -8222,20 +8234,16 @@ H5D__chunk_iter(H5D_t *dset, H5D_chunk_iter_op_t op, void *op_data)
 
 /*-------------------------------------------------------------------------
- * Function:    H5D__chunk_get_offset_copy
+ * Function:    H5D__chunk_verify_offset
```
Turns out nothing in the direct chunk code modifes the offset array so there's no need to copy it and we can just pass it directly from the user qualified as const. Added this function to perform verification.

fortnern commented on src/H5Dio.c:
```diff
@@ -302,22 +315,25 @@ H5D__read(size_t count, H5D_dset_io_info_t *dset_info)
-    /* Perform second phase of type info initialization */
-    if (H5D__typeinfo_init_phase2(&io_info) < 0)
-        HGOTO_ERROR(H5E_DATASET, H5E_CANTINIT, FAIL, "unable to set up type info (second phase)");
+    /* Initialize type info differently depending on if weŕe using the shared chunk cache or not */
```
The shared chunk cache will need to perform its own type conversion initialization.

fortnern commented on src/H5Dpkg.h:
```diff
@@ -115,6 +115,13 @@
 /****************************/
 
+/* Typedef for cached dataset creation property list information */
+typedef struct H5D_dcpl_cache_t {
```
This is not used outside of the H5D package so I demoted it to this header to minimize issues with circular dependencies

fortnern commented on src/H5VLnative_dataset.c:
```diff
@@ -755,53 +756,75 @@ H5VL__native_dataset_optional(void *obj, H5VL_optional_args_t *args, hid_t dxpl_
-                0)
-                HGOTO_ERROR(H5E_DATASET, H5E_READERROR, FAIL, "can't read unprocessed chunk data");
+                /* Read the raw chunk */
+                if (H5D__chunk_direct_read(dset, chunk_read_args->offset, &chunk_read_args->filters,
```
Checking the validity of the offset is layout specific and has been moved to H5Dchunk.c for the chunked non-SCC case.

fortnern commented on test/struct_chunk_storage.c:
```diff
@@ -235,6 +235,9 @@ test_sparse_direct_chunk(hid_t fapl)
 
     TESTING("APIs for direct chunk I/O on structured chunks");
 
+    SKIPPED();
```
Skipped since struct chunk direct I/O requires a shared chunk cache client implementation, which we don't have yet.

vchoi-hdfgroup commented on src/H5SCprivate.h:
```diff
@@ -58,71 +61,79 @@ typedef herr_t (*H5SC_chunk_lookup_t)(H5D_t *dset, size_t count, hsize_t *scaled
  * of the chunk buffer is returned in *buf_size. */
 typedef herr_t (*H5SC_new_chunk_t)(H5D_t *dset, bool fill, size_t *nbytes /*out*/, size_t *buf_size /*out*/,
-                                   void **chunk /*chunk*/);
+                                   void **chunk /*chunk*/, void **udata[] /*out*/);
```
Is the udata[] correct here?

fortnern replied:
Yes, because it's an array of udata pointers, one for each chunk

vchoi-hdfgroup commented on src/H5SCprivate.h:
```diff
@@ -58,71 +61,79 @@ typedef herr_t (*H5SC_chunk_lookup_t)(H5D_t *dset, size_t count, hsize_t *scaled
 typedef herr_t (*H5SC_chunk_insert_t)(H5D_t *dset, size_t count, hsize_t *scaled[] /*in*/,
                                       haddr_t *addr[] /*in,out*/, hsize_t old_disk_size[],
-                                      hsize_t new_disk_size[], void *chunk[] /*in*/);
+                                      hsize_t new_disk_size[], void *chunk[] /*in*/, void *udata);
```
Should the udata be an array?

fortnern replied:
Looks like I vectorized the wrong udata, thanks!

fortnern committed:
Fix layout callback signatures and add chunk_dims parameter to
layout_query

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Integrate shared chunk cache with get_defined and erase operations.
Various other fixes/enhancements.

fortnern committed:
Merge branch 'feature/sparse_data' of github.com:fortnern/hdf5 into feature/sparse_data

github-actions[bot] committed:
Committing clang-format changes

fortnern committed:
Fix failing struct_chunk_storage test, since H5Derase() throws an error
when it isn't supported. Also replace inappropriate usage of
FAIL_STACK_ERROR with TEST_ERROR.

qkoziol commented:
@fortnern - Is this ready (enough :-) to review?

fortnern commented:
> @fortnern - Is this ready (enough :-) to review?

You're welcome to review it if you like. It's nowhere near production ready yet though - just part of the machinery for the shared chunk cache (designed to support sparse chunks), which hasn't been implemented yet. Note the PR is to a feature branch not develop.

jhendersonHDF commented on src/H5Dio.c:
```diff
@@ -286,10 +289,20 @@ H5D__read(size_t count, H5D_dset_io_info_t *dset_info)
-            (dset_info[i].layout_ops.io_init)(&io_info, &(dset_info[i])) < 0)
-            HGOTO_ERROR(H5E_DATASET, H5E_CANTINIT, FAIL, "can't initialize I/O info");
+        /* Check for shared chunk cache support */
+        if (dset_info[i].dset->shared->layout.sc_ops)
```
What are the current cases where a dataset will not support the shared chunk cache? It would be fantastic if we could eliminate large amounts of the code that falls in the other path here.

fortnern replied:
Indeed, the idea is to eventually have every dataset type support the shared chunk cache. Currently none of them do, though structured chunk will support it from the beginning

jhendersonHDF commented on src/H5SC.c:
```diff
@@ -155,21 +157,24 @@ H5SC_flush_dset(H5SC_t *cache, H5D_t *dset, bool evict)
+ * datasets in the dset_info array that do not support the shared chunk
+ * cache. These datasets must be ignored by the shared chunk cache. There
+ * may also be datasets that have skip_io set. These datasets must also be
+ * skipped.
```
While it may look a bit awkward at the higher levels, I propose that we partition up the `H5D_dset_io_info_t` structs into separate arrays (or do a sort) based on those that support H5SC and those which don't (also the same for the `skip_io` case) to simplify the logic here and not push the concerns of skipping over datasets down to these functions.

fortnern replied:
We could certainly do that. My initial idea is that it won't be difficult or costly to just skip over the unused datasets (as is the case in H5Dio.c, where it skips over datasets that DO support SCC), and is not worth the overhead, but if it ends up being difficult we can create a new list.

jhendersonHDF commented on src/H5Dchunk.c:
```diff
@@ -8260,70 +8262,11 @@ H5D__chunk_get_offset_copy(const H5D_t *dset, const hsize_t *offset, hsize_t *of
- *-------------------------------------------------------------------------
- */
-herr_t
-H5D__write_struct_chunk_direct(H5D_t H5_ATTR_UNUSED *dset, hsize_t H5_ATTR_UNUSED *offset,
```
Unused at this point or were these moved elsewhere?

fortnern replied:
Unused, should not be in H5Dchunk.c, and should be implemented in terms of H5SC layout callbacks. The H5SC code will handle the actual direct I/O, the only thing the layout callbacks need to do is insert the chunk into the index.

fortnern committed:
Adjust H5D_io_info_t struct and add more descriptions for the fields.

github-actions[bot] committed:
Committing clang-format changes
